src/main/scala/org/ergoplatform/validation/ValidationRules.scala

```diff
@@ -163,6 +180,51 @@ object ValidationRules {
     }
   }
 
+  /** For CheckCostFuncOperation we use 1-511 range op codes. Thus
+   * `ChangedRule.newValue` should be parsed as a sequence of `getUShort`
+   * values and then the exOpCode should be checked against that parsed
+   * sequence.
+   * Note, we don't need to store a number of items in a sequence,
+   * because at the time of parsing we may assume that `ChangedRule.newValue`
+   * has correct length, so we just parse it until end of bytes (of cause
+   * checking consistency). */
+  object CheckCostFuncOperation extends ValidationRule(1014,
+    "Check the opcode is allowed in cost function") with SoftForkWhenCodeAdded {
+    def apply[Ctx <: IRContext, T](ctx: Ctx)(opCode: OpCodeExtra)(block: => T): T = {
+      def msg = s"Not allowed opCode $opCode in cost function"
+      def args = Seq(unsignedOpCode(opCode))
+      validate(ctx.isAllowedOpCodeInCosting(opCode), new CosterException(msg, None), args, block)
+    }
+
+    override def isSoftFork(vs: SigmaValidationSettings,
+                            ruleId: Short,
+                            status: RuleStatus, args: Seq[Any]): Boolean = (status, args) match {
+      case (ChangedRule(newValue), Seq(code: OpCodeExtra)) =>
+        decodeVLQUShort(newValue).contains(code)
+      case _ => false
+    }
+
+    @inline
+    private def unsignedOpCode(opCode: OpCodeExtra): Short =
+      if (opCode < 0) opCode.toByte.toUByte.toShort else opCode
```

Comment:
@greenhat Could you please explain what this means? OpCodeExtra values should never be negative.

Reply:
Just as I was writing an explanation I found a way to eliminate it. :) Here is the fix - https://github.com/ScorexFoundation/sigmastate-interpreter/pull/503/commits/ccbdd58c50e4e205882f8f446c243a9d76ac887d